e2e/pos_test.go

```diff
@@ -0,0 +1,617 @@
+	numTransactions := 6
+	signer := crypto.NewEIP155Signer(100)
+	currentNonce := 0
+
+	// TxPool client
+	txpoolClient := srv.TxnPoolOperator()
+
+	generateTx := func(value *big.Int, methodName string) *types.Transaction {
+		signedTx, signErr := signer.SignTx(&types.Transaction{
+			Nonce:    uint64(currentNonce),
+			From:     types.ZeroAddress,
+			To:       &stakingContractAddr,
+			GasPrice: bigGasPrice,
+			Gas:      framework.DefaultGasLimit,
+			Value:    value,
+			V:        big.NewInt(1), // it is necessary to encode in rlp
+			Input:    framework.MethodSig(methodName),
+		}, senderKey)
+
+		if signErr != nil {
+			t.Fatalf("Unable to sign transaction, %v", signErr)
+		}
+
+		currentNonce++
+
+		return signedTx
+	}
+
+	oneEth := framework.EthToWei(1)
+	zeroEth := framework.EthToWei(0)
+	for i := 0; i < numTransactions; i++ {
+		var msg *txpoolOp.AddTxnReq
+		if i%2 == 0 {
+			unstakeTxn := generateTx(zeroEth, "unstake")
+			msg = &txpoolOp.AddTxnReq{
+				Raw: &any.Any{
+					Value: unstakeTxn.MarshalRLP(),
+				},
+				From: types.ZeroAddress.String(),
+			}
+		} else {
+			stakeTxn := generateTx(oneEth, "stake")
+			msg = &txpoolOp.AddTxnReq{
+				Raw: &any.Any{
+					Value: stakeTxn.MarshalRLP(),
+				},
+				From: types.ZeroAddress.String(),
+			}
+		}
+
+		_, addErr := txpoolClient.AddTxn(context.Background(), msg)
+		if addErr != nil {
+			t.Fatalf("Unable to add txn, %v", addErr)
+		}
+	}
+
+	// Set up the blockchain listener to catch the added block event
+	blockNum := waitForBlock(t, srv, 1, 0)
+
+	block, blockErr := client.Eth().GetBlockByNumber(web3.BlockNumber(blockNum), true)
+	if blockErr != nil {
+		t.Fatalf("Unable to fetch block")
+	}
+
+	// Find how much the account paid for all the transactions in this block
+	paidFee := big.NewInt(0).Mul(bigGasPrice, big.NewInt(int64(block.GasUsed)))
+
+	// Check the balances
+	actualAccountBalance := framework.GetAccountBalance(senderAddr, client, t)
+	actualStakingSCBalance, fetchError := framework.GetStakedAmount(senderAddr, client)
+	if fetchError != nil {
+		t.Fatalf("Unable to fetch staking SC balance, %v", fetchError)
+	}
+
+	expStake := big.NewInt(0).Mul(big.NewInt(int64(numDummyStakers)), bigDefaultStakedBalance)
+	expStake.Add(expStake, oneEth)
+
+	assert.Equalf(t,
+		expStake.String(),
+		actualStakingSCBalance.String(),
+		"Staked address balance mismatch after stake / unstake exploit",
+	)
+
+	// Make sure the account balances match up
+
+	// expBalance = previousAccountBalance + stakeRefund - 1 ETH - block fees
+	expBalance := big.NewInt(0).Sub(big.NewInt(0).Add(defaultBalance, bigDefaultStakedBalance), oneEth)
+	expBalance = big.NewInt(0).Sub(expBalance, paidFee)
+
+	assert.Equalf(t,
+		expBalance.String(),
+		actualAccountBalance.String(),
+		"Account balance mismatch after stake / unstake exploit",
+	)
+}
+
+// Test scenario:
+// User has 0 ETH staked and a balance of 100 ETH
+// Stake 2 ETH -> Unstake
+// Expected result for tests: Staked: 0 ETH; Balance: ~100 ETH
```

Comment:
Also add this check, please:
```suggestion
// Expected result for tests: Staked: 0 ETH; Balance: ~100 ETH; not a validator
```

Reply:
Fixed 👍 

28f5205